File: backend/app/src/main/kotlin/net/animeta/backend/controller/v2/UserRecordsController.kt
package net.animeta.backend.controller.v2

import net.animeta.backend.db.Datastore
import net.animeta.backend.db.query
import net.animeta.backend.dto.RecordDTO
import net.animeta.backend.exception.ApiException
import net.animeta.backend.model.QRecord.record
import net.animeta.backend.model.StatusType
import net.animeta.backend.model.User
import net.animeta.backend.repository.HistoryRepository
import net.animeta.backend.repository.RecordRepository
import net.animeta.backend.repository.UserRepository
import net.animeta.backend.security.CurrentUser
import net.animeta.backend.serializer.RecordSerializer
import net.animeta.backend.service.UserRecordsService
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.PathVariable
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RequestParam
import org.springframework.web.bind.annotation.RestController
import java.util.Optional

@RestController
@RequestMapping("/v2/users/{name}/records")
class UserRecordsController(val userRepository: UserRepository,
                            val recordRepository: RecordRepository,
                            val historyRepository: HistoryRepository,
                            val userRecordsService: UserRecordsService,
                            val datastore: Datastore,
                            val recordSerializer: RecordSerializer) {
    data class GetWithCountsResponse(val data: List<RecordDTO>, val counts: UserRecordsService.RecordCounts)

    @GetMapping
    fun get(@PathVariable name: String,
            @CurrentUser(required = false) currentUser: User?,
            @RequestParam("include_has_newer_episode", defaultValue = "false") includeHasNewerEpisodeParam: Boolean,
            @RequestParam("status_type", required = false) statusTypeParam: String?,
            @RequestParam("category_id", required = false) categoryIdParam: Int?,
            @RequestParam(required = false) sort: String?,
            @RequestParam(required = false) limit: Int?,
            @RequestParam("with_counts", defaultValue = "false") withCounts: Boolean,
            @RequestParam(required = false) options: RecordSerializer.Options?): Any {
        val user = userRepository.findByUsername(name) ?: throw ApiException.notFound()
        val includeHasNewerEpisode = includeHasNewerEpisodeParam && currentUser?.id == user.id
        val statusType = if (statusTypeParam != null && statusTypeParam != "") {
            StatusType.valueOf(statusTypeParam.toUpperCase())
        } else {
            null
        }
        var query = record.query.filter(record.user.eq(user))
        if (statusType != null) {
            query = query.filter(record.status_type.eq(statusType))
        }
        val categoryId: Optional<Int>? = categoryIdParam?.let { Optional.of(it) }?.filter { it != 0 }
        if (categoryId != null) {
            query = query.filter(categoryId.map { record.category.id.eq(it) }.orElse(record.category.isNull))
        }
        when (sort) {
            "date", null -> query = query.orderBy(record.updated_at.desc())
            "title" -> query = query.orderBy(record.title.asc())
        }
        if (limit != null) {
            query = query.limit(limit)
        }
        val data = datastore.query(query).map {
            recordSerializer.serialize(it, options ?: RecordSerializer.legacyOptions(includeHasNewerEpisode = includeHasNewerEpisode))
        }
        if (!withCounts) {
            return data
        }
        return GetWithCountsResponse(data, userRecordsService.count(user, statusType, categoryId))
    }
}

File: backend-ts/src/controllers/user_records.controller.ts
import { Controller, DefaultValuePipe, Get, Param, ParseBoolPipe, Query } from "@nestjs/common";
import { PostDTO, RecordDTO } from 'shared/types';
import { ApiException } from "./exceptions";
import { LessThan, Repository } from "typeorm";
import { InjectRepository } from "@nestjs/typeorm";
import { PostSerializer, PostSerializerOptions } from "src/serializers/post.serializer";
import { History } from 'src/entities/history.entity'
import { User } from "src/entities/user.entity";
import { ParseIntOrNullPipe } from "src/utils/parse_int_or_null.pipe";
import { RecordService } from "src/services/record.service";
import { RecordsCount, UserRecordsService } from "src/services/user_records.service";
import { CurrentUser } from "src/auth/decorators";
import { StatusType } from "src/entities/status_type";
import { Record } from "src/entities/record.entity";
import { RecordSerializer, RecordSerializerOptions } from "src/serializers/record.serializer";

type GetWithCountsResponse = {
  data: RecordDTO[];
  counts: RecordsCount;
}

@Controller('/api/v4/users/:name/records')
export class UserRecordsController {
  constructor(
    private userRecordsService: UserRecordsService,
    @InjectRepository(User) private userRepository: Repository<User>,
    @InjectRepository(Record) private recordRepository: Repository<Record>,
    private recordSerializer: RecordSerializer,
  ) {}

  @Get()
  async get(
    @Param('name') name: string,
    @CurrentUser({ required: false }) currentUser: User | null,
    @Query('include_has_newer_episode', new DefaultValuePipe('false'), new ParseBoolPipe()) includeHasNewerEpisodeParam: boolean,
    @Query('status_type') statusTypeParam: string | null,
    @Query('category_id', new ParseIntOrNullPipe()) categoryIdParam: number | null,
    @Query('sort') sort: string | null,
    @Query('limit', new ParseIntOrNullPipe()) limit: number | null,
    @Query('with_counts', new DefaultValuePipe('false'), new ParseBoolPipe()) withCounts: boolean,
    @Query('options') optionsJson: string | null,
  ): Promise<RecordDTO[] | GetWithCountsResponse> {
    const user = await this.userRepository.findOne({ where: {username: name} })
    if (!user) throw ApiException.notFound()
    const includeHasNewerEpisode = includeHasNewerEpisodeParam && currentUser?.id === user.id
    let statusType: StatusType | null = null
    if (statusTypeParam != null && statusTypeParam !== '') {
      statusType = StatusType[statusTypeParam.toUpperCase() as keyof typeof StatusType]
    }
    const records = await this.recordRepository.find({
      where: {
        user,
        ...statusType != null ? { status_type: statusType } : {},
        ...categoryIdParam != null ? {
          category_id: categoryIdParam !== 0 ? categoryIdParam : null
        } : {},
      },
      order: sort === 'title' ? { title: 'ASC' } :
        sort === 'rating' ? { rating: 'DESC', updated_at: 'DESC' } :
        /* sort === 'date' || !sort */ { updated_at: 'DESC' },
      ...limit ? { take: limit } : {},
    })
    const options: RecordSerializerOptions =
      optionsJson ? JSON.parse(optionsJson) : RecordSerializer.legacyOptions({
        includeHasNewerEpisode,
      })
    const data = await Promise.all(records.map(it => this.recordSerializer.serialize(it, options)))
    if (!withCounts) {
      return data
    }
    return {
      data,
      counts: await this.userRecordsService.count(user, statusType, categoryIdParam),
    }
  }
}
